spartan_parallel/src/scalar/fp.rs

```diff
@@ -0,0 +1,204 @@
+use super::SpartanExtensionField;
+use crate::{AppendToTranscript, ProofTranscript, Transcript};
+use ceno_goldilocks::Goldilocks;
+use core::borrow::Borrow;
+use core::iter::{Product, Sum};
+use core::ops::{Add, AddAssign, Mul, MulAssign, Sub, SubAssign};
+use ff::{Field, FromUniformBytes};
+use rand::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
+use std::ops::Neg;
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq, CtOption};
+use zeroize::Zeroize;
+
+/// Constant representing the modulus
+/// q = 2^64 - 2^32 + 1
+/// 0xFFFFFFFF00000001
+const P: u64 = 0xFFFF_FFFF_0000_0001;
+
+/// Field wrapper around base Goldilocks
+#[derive(Clone, Copy, Eq, Serialize, Deserialize, Hash, Debug)]
+pub struct Scalar(Goldilocks);
+
+impl SpartanExtensionField for Scalar {
+  type InnerType = Goldilocks;
+
+  fn inner(&self) -> &Goldilocks {
+    &self.0
+  }
+
+  fn field_zero() -> Self {
+    Goldilocks::ZERO.into()
+  }
+
+  fn field_one() -> Self {
+    Goldilocks::ONE.into()
+  }
+
+  fn random<Rng: RngCore + CryptoRng>(rng: &mut Rng) -> Self {
```

Comment:
I think you can take that from the Goldilocks crate.  (If not, please add it there, and then take it from there.)

Reply:
Updated.